### ISSUE TITLE
Security issue: JDBC URI contains plaintext password & username

### ISSUE THREAD

**Issue by roman-novosad** [AbsaOSS/spline-spark-agent#69]
## Proposal
Spline Agent should contain URI filtering for, at least, passwords. Captured Lineage data are send insecurely thru HTTP protocol to Producer, which can be located out of the secured network, so anyone can listen and capture the password. 

I believe this issue should be solved on either Agent and UI sides. Best solution is not sent it at all or securely mask it (f.e. with asterisks). 

## Example
Current state:
`DataSource URI: jdbc:sqlserver://sample.database.windows.net:1433;database=sample;user=sample;password=password;encrypt=true;trustServerCertificate=false;hostNameInCertificate=*.database.windows.net;loginTimeout=30:
`

Future state:
`DataSource URI: jdbc:sqlserver://sample.database.windows.net:1433;database=sample;user=;password=;encrypt=true;trustServerCertificate=false;hostNameInCertificate=*.database.windows.net;loginTimeout=30:
`
## Version 
For version 0.5.1 Agent, UI.

**Comment by radford1** [comment]
This might be a little difficult to solve as a blanket fix. Most drivers have the variable as password=<password> so that would be easy to catch and replace with **** but I'm not sure if all jdbc drivers have this exact parameter name as password=<password>. Would it be ok to implement a small method that will take the url and start by replacing password=<password> with password=****. If new variations of password come up we can implement more logic in that method.

**Comment by wajda** [comment]
Very good point. JDBC standard is indeed too flexible in terms ways to pass credentials to the driver.
Here's a related StackOverflow question - https://stackoverflow.com/questions/5547224/username-password-in-jdbc-connection-url

The problem here is a bit more general than just hiding JDBC passwords.
Ideally a data source URI should only contain parts that identify the location (no auth credentials, no other extra connection parameters, protocol variations etc). Because this is the primary goal for the datasource URI - to identify (uniquely and consistently) a data location in the network. Unfortunately many storage providers use URLs or connection strings to not only identify a location but also pass a bunch of other auth and configuration stuff to the driver.

We plan to implement a mechanism on the Spline server side to manage the data source info, merge logical duplicates, give them display names etc. But on the agent it would be cool to go as far as we can trying to recognize and parse the connection strings, using a knowledge about a specific data source type in use, and reconstruct the Spline datasource URI using only necessary data (like server/port, db name, table name, etc). This would solve the password issue automatically. We could however store an original connection string as an extra write/read operation parameter, just for the record purposes, and make that option configurable.
But for the source types that have too flexible connection string/url format, so we cannot reliably parse, we have to save them verbatim, and they could leak passwords anyway.

**Comment by radford1** [comment]
Ok sounds good. I'll put something together along these lines in the next week or two and we can review

**Comment by wajda** [comment]
Awesome. Thank you David!

**Comment by ASchwad** [comment]
Sounds like a good feature. #139 is merged by now, could you share some insights how to leverage the feature by adding an example to the docs?

**Comment by cerveada** [comment]
This is a part of Spline 0.6 that is not released yet. We will create docs for it when it's ready.

But the usage is simple. An Example of custom filter is [DataSourceUriLineageFilter](https://github.com/AbsaOSS/spline-spark-agent/blob/develop/core/src/main/scala/za/co/absa/spline/harvester/postprocessing/DataSourceUriLineageFilter.scala). You can create custom filters the same way by extending AbstractLineageFilter.

Then you just set the full class name of the filter in properties:
```
 spline.postprocessing_filter.classNames=foo.bar.MyFilter
```
Some documentation of properties is in [spline.default.properties](https://github.com/AbsaOSS/spline-spark-agent/blob/develop/core/src/main/resources/spline.default.properties).

**Comment by ASchwad** [comment]
Please correct me if I am wrong: when I want to use the DataSourceUriLineageFilter, I should add the following line into the spline.default.properties, correct? (for testing purposes, even if 0.6 is not yet released)

`spline.postprocessing_filter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter`

**Comment by cerveada** [comment]
That will work, but usually the properties are set externally, so you can change them without changing the jar

https://absaoss.github.io/spline/#properties

`spline.default.properties` are meant to be a fallback when the user does not set the property himself.

**Comment by araujo-filipe** [comment]
Hey, I've been trying to use this feature using the code from the develop branch, but it doesn't seem to be working...
I've tried setting either by themselves or together these properties:
spline.postprocessingFilter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter
spline.postprocessing_filter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter

But the password always shows up in the execution plan. I've even tried to play around with the regex of this filter a bit but with no progress. Any idea of what I can be doing wrong?

(I know this feature is not yet released, but it is necessary for the development I'm trying to do.)

**Comment by wajda** [comment]
@araujo-joao,
This one was correct in your snapshot `spline.postprocessingFilter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter`.
But I just created a pull-request that changes the way filters are configured, as well as the name of the filter.
From now on you can enable the pre-configured password replacing filter like this:
```properties
# explicitly specify the root post processing filter
spline.postProcessingFilter=dsUriPasswordReplace

# ... or, add it to the composite filter sub-filters list if you want multiple other filters to be executed as well
spline.postProcessingFilter=composite
spline.postProcessingFilter.composite.filters=dsUriPasswordReplace, ... other filter names
```
See this change for details - https://github.com/AbsaOSS/spline-spark-agent/pull/213/files#diff-baadc4144b8729b5fd1816ededaead9a68f7d17a72eb2a1539cb774577325305

**Comment by wajda** [comment]
But if the previous config that you tries didn't work, there might be a bug. Can you share the URI example where passwords aren't replaced by the filter?
Also note that the filter only processes data source URIs stored in the `ReadOperation.inputSources` and `WriteOperation.outputSource`. It doesn't scan other operation properties.

**Comment by araujo-filipe** [comment]
Hey @wajda, thanks for the replies. After looking a bit into the code (scala noob) I figured out exactly what you said on the last reply... I am giving spark the password as an option to the reader and not on the jdbc url itself, so the password shows up in  `ReadOperation.params` and is not filtered.
 
**Py-spark:**
```
df = spark_session.read.format("jdbc")
    .option("url", "jdbc:postgresql://someHost:somePort/someDB")
    .option("dbtable", "someTable")
    .option("user", "someUser")
    .option("password", "somePassword")
    .load()
```
**Returned execution plan:**
```
"reads": [
            {
                "inputSources": [
                    "jdbc:postgresql://someHost:somePort/someDB:someTable"
                ],
                "params": {
                    "url": "jdbc:postgresql://someHost:somePort/someDB",
                    "dbtable": "someTable",
                    "user": "someUser",
                    "password": "somePassword"
                },
               ...
        ],
```


One thing I noticed was that even if I do feed the password *within* the jdbc url instead of an option to the reader, although it is filtered on the `ReadOperation.inputSources`, it still shows up unfiltered on the `ReadOperation.params`.

**Py-spark:**
```
df = spark_session.read.format("jdbc")
    .option("url", "jdbc:postgresql://someHost:somePort/someDB?user=someUser&password=somePassword")
    .option("dbtable", "someTable")
    .load()
```
**Returned execution plan:**
```
"reads": [
            {
                "inputSources": [
                    "jdbc:postgresql://someHost:somePort/someDB?user=someUser&password=*****"
                ],
                "params": {
                    "url": "jdbc:postgresql://someHost:somePort/someDB?user=someUser&password=somePassword",
                    "dbtable": "someTable",
                },
               ...
        ],
```

So, in conclusion, I understand that the feature was only designed to filter the password from `ReadOperation.inputSources` but if it still shows up in `ReadOperation.params` it won't solve the security issue. Maybe this is just happening on my case, not sure.

**Comment by wajda** [comment]
Yeah, we need to scan for passwords in more places. Will create a PR

**Comment by wajda** [comment]
In addition to the `inputSources` and `outputSource` properties, the filter now also looks for sensitive values in `params` of the `Read` and `Write` operations. (The other operations are _not_ scanned as we assume no secrets in data transformation operations).

**Comment by xfontes42** [comment]
Hello @wajda , looking at the most recent changes and the PR for this issue, cases like this are completely solved:
```json
"reads": [
{
"inputSources": [
"jdbc:postgresql://someHost:somePort/someDB?user=someUser&password=*****"
],
"params": {
"url": "jdbc:postgresql://someHost:somePort/someDB?user=someUser&password=******",
"dbtable": "someTable",
},
```
However, there is still the case:
```json
"params": {
"url": "jdbc:postgresql://someHost:somePort/someDB",
"dbtable": "someTable",
"user": "someUser",
"password": "somePassword"
}
```
Would it be possible to change the regex in the properties such that this scenario is also handled?
Thanks in advance for your help!

**Comment by wajda** [comment]
Are you looking at the pull request - #220? It's not yet merged.
If so then it would be strange that it's not working, as there is a unit test that specifically checks the pattern you mentioned - https://github.com/AbsaOSS/spline-spark-agent/pull/220/files#diff-def5fcded193d2e425683f495ee3bdfe97a890043c2e6182b74399183700a34fR85

The filter uses two sets of regular expressions - one is applied on keys (property names), and the other - on the values.
See - https://github.com/AbsaOSS/spline-spark-agent/pull/220/files#diff-baadc4144b8729b5fd1816ededaead9a68f7d17a72eb2a1539cb774577325305R100

Check if you don't have those defaults overridden in your conf.

**Comment by xfontes42** [comment]
Hello @wajda ,

I've checked and I wasn't overriding the defaults.
I also tested adding to the `spline.postProcessingFilter.dsPasswordReplace.nameRegexes` property the specific name for my case (basically just `password`), still to no avail.

It might also be a particular circumstance of my setup, so I will look a bit further into that. 
Nonetheless, I appreciate your quick answer!

**Comment by wajda** [comment]
Are those "params" located in a read/write operation or a generic (other) operation?

**Comment by xfontes42** [comment]
They're in a read operation.

**Comment by wajda** [comment]
I got it, nevermind :facepalm:  will fix shortly :)

**Comment by wajda** [comment]
Please check it now, should be working.
It was my oversight, the filtering was applied for the n-th level of nested properties, but not the 1st one.
I've added more unit tests as well.

**Comment by xfontes42** [comment]
Works perfectly! 
Thank you @wajda
